### PR TITLE
feat: generate configMeta

### PR DESCRIPTION
**Pull request by popomore**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

config

##### Description of change

The configMeta has config structure, but the value is the filepath
which defines the config

### CODE REVIEW

**Comment by popomore** [comment]
https://github.com/eggjs/egg/issues/1132

File: lib/loader/mixin/config.js [hunk 1]
```diff
@@ -90,7 +90,29 @@ module.exports = {
...
+  [SET_CONFIG_META](config, filepath) {
+    config = extend(true, {}, config);
+    setConfig(config, filepath);
+    extend(true, this.configMeta, config);
+  },
 };
+
+function setConfig(obj, filepath) {
+  for (const key of Object.keys(obj)) {
+    if (isObject(obj[key])) {
```
Comment (fengmk2) [inline lib/loader/mixin/config.js]: 如果是 httpclient，configclient 这种 value 呢？会深度遍历它么？

File: test/fixtures/configmeta/config/config.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+'use strict';
+
+exports.urllib = {
+  keepAlive: false
+};
```
Comment (fengmk2) [inline test/fixtures/configmeta/config/config.js]: 加一个 `exports.foo = { httpclient: urllib, logger: console }` 看看会怎样？

Comment (popomore) [inline test/fixtures/configmeta/config/config.js]: urllib 无法判断，他就是一个 {}

File: lib/loader/mixin/config.js [hunk 3]
```diff
@@ -90,7 +90,25 @@ module.exports = {
...
+  [SET_CONFIG_META](config, filepath) {
+    config = extend(true, {}, config);
+    setConfig(config, filepath);
+    extend(true, this.configMeta, config);
+  },
 };
+
+function setConfig(obj, filepath) {
+  for (const key of Object.keys(obj)) {
+    if (Object.getPrototypeOf(obj[key]) === Object.prototype) {
```
Comment (popomore) [inline lib/loader/mixin/config.js]: 这里原来确实有 bug

**Comment by popomore** [comment]
@fengmk2 再看看

**Comment by fengmk2** [comment]
@popomore 没问题了，你来合并发布吧。

**Comment by popomore** [comment]
+ egg-core@3.12.0

File: lib/loader/mixin/config.js (hunk 3 above)
Comment (fengmk2) [inline lib/loader/mixin/config.js]: obj[key] 为 null 的时候有 bug

image

**Pull request merged**